### PR TITLE
feat: allow prefixes at the beginning of subdomain app hostnames

### LINKED ISSUE #9903: feature request: subdomain on applications
We need to support urls in the form of `http://subdomain.3000--main--test3--admin.coder.com` to be forwarded to the application, i started playing around with the required changes, some pointers would be great! (and how to test it too)

https://github.com/coder/coder/compare/main...raphaelfff:coder:patch-1

### PR DESCRIPTION
**Pull request by deansheather**
Changes hostname parsing and signed app token generation/verification to allow for `(prefix---)?app--agent--workspace--user` in subdomain app URLs. Signed app tokens now include the prefix in their payload and will verify it on new requests.

~~Old signed app tokens will become invalid (but they have an expiry of 1 minute anyways so it doesn't matter).~~ They actually won't become invalid because prefixes weren't a thing before so the JSON parsing will default to `""`.

The prefix must be separated by 3 hyphens to distinguish it from the rest of the segments in the URL. This is to allow for changes to our URL format (like adding or removing segments) in the future without conflicting with prefixes.

E.g. `https://frontend---8080--main--dev--deansheather.example.com` is the same app as `https://8080--main--dev--deansheather.example.com`.

Apps can view the full Host so they can route based on this prefix if they wish.

## TODO:
- [ ] ~~Put behind a configuration flag (disabled by default)~~

Closes #9903

### CODE REVIEW

**Comment by Emyrk** [comment]
Do we need to have it toggleable?

File: coderd/httpapi/url.go [hunk 1]
```diff
@@ -133,8 +157,12 @@ func CompileHostnamePattern(pattern string) (*regexp.Regexp, error) {
 	}
 	for i, label := range strings.Split(pattern, ".") {
 		if i == 0 {
-			// We have to allow the asterisk to be a valid hostname label.
+			// We have to allow the asterisk to be a valid hostname label, so
+			// we strip the asterisk (which is only on the first one).
 			label = strings.TrimPrefix(label, "*")
+			// Put an "a" at the start to stand in for the asterisk in the regex
+			// test below. This makes `*.coder.com` become `a.coder.com` and
+			// `*--prod.coder.com` become `a--prod.coder.com`.
```
Comment (Emyrk) [inline coderd/httpapi/url.go:165]: :+1:

File: coderd/workspaceapps/apptest/setup.go [hunk 2]
```diff
@@ -25,13 +25,14 @@ import (
...
 	"github.com/coder/coder/v2/codersdk/agentsdk"
+	"github.com/coder/coder/v2/cryptorand"
 	"github.com/coder/coder/v2/provisioner/echo"
 	"github.com/coder/coder/v2/provisionersdk/proto"
 	"github.com/coder/coder/v2/testutil"
 )
 
 const (
-	proxyTestAgentName            = "agent-name"
+	proxyTestAgentName            = "agnt-name"
```
Comment (Emyrk) [inline coderd/workspaceapps/apptest/setup.go:35]: Lol did we hit a limit in the test?

**Comment by deansheather** [comment]
I guess we don't have to

File: coderd/workspaceapps/request.go [hunk 3]
```diff
@@ -170,6 +174,13 @@ func (r Request) Validate() error {
 		return xerrors.New("app slug or port is required")
 	}
 
+	if r.Prefix != "" && r.AccessMethod != AccessMethodSubdomain {
+		return xerrors.New("prefix is only valid for subdomain apps")
+	}
```
Comment (Emyrk) [inline coderd/workspaceapps/request.go:179]: Smart

File: coderd/httpapi/url.go [hunk 4]
```diff
@@ -55,20 +57,42 @@ func (a ApplicationURL) String() string {
...
+// reduce the character count for app URLs.
 func ParseSubdomainAppURL(subdomain string) (ApplicationURL, error) {
+	var (
+		prefixSegments = strings.Split(subdomain, "---")
+		prefix         = ""
+	)
+	if len(prefixSegments) > 1 {
+		prefix = strings.Join(prefixSegments[:len(prefixSegments)-1], "---") + "---"
+		subdomain = prefixSegments[len(prefixSegments)-1]
+	}
```
Comment (Emyrk) [inline coderd/httpapi/url.go:86]: I guess someone could `a---b---<rest>`, I would have only split the once.

**Comment by MrPeacockNLB** [comment]
Hi @deansheather,

does "we can change URL format later" mean you can fix this #9186 ? We are planing to deploy a combination of WAP + WAF in front of Coder. So this mean we have to disable this standard OWASP rule in order to get this working.

**Comment by deansheather** [comment]
> does "we can change URL format later" mean you can fix this https://github.com/coder/coder/issues/9186 ? We are planing to deploy a combination of WAP + WAF in front of Coder. So this mean we have to disable this standard OWASP rule in order to get this working.

@MrPeacockNLB we don't have any plans to fix that issue. We use double hyphens so we can support single hyphens in app names, agent names, workspace names and usernames.

**Comment by deansheather** [comment]
The "we can change the URL format later" is more about adding or removing segments. The agent name segment is not necessary if you provide an app slug because apps are unique per template, not per agent. Currently the URL format requires the agent name, but in the future we could make it not required.

File: coderd/workspaceapps/apptest/setup.go (hunk 2 above)
Comment (deansheather) [reply coderd/workspaceapps/apptest/setup.go:35]: Yeah we hit the 62 char limit lol, Marcin reverted his PR though so it should be fine now hopefully

File: coderd/httpapi/url.go (hunk 4 above)
Comment (deansheather) [reply coderd/httpapi/url.go:86]: yeah that should result in `Prefix: "a---b---"`, which is what we want.

**Review by Emyrk — APPROVED** [review]
Up to you if you want to make it a deployment toggle

**Pull request merged**